### ISSUE TITLE
[Bug] Trained SimCLR model shows a different accuracy in linear probing with each run

### ISSUE THREAD

**Issue by guneetmutreja** [open-mmlab/mmselfsup#724]
### Branch

1.x branch (1.x version, such as `v1.0.0rc2`, or `dev-1.x` branch)

### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmselfsup/issues) and [Discussions](https://github.com/open-mmlab/mmselfsup/discussions) but cannot get the expected help.
- [X] I have read the [documentation](https://mmselfsup.readthedocs.io/en/latest/) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version](https://github.com/open-mmlab/mmselfsup).

### Environment

sys.platform: linux
Python: 3.8.16 (default, Jan 17 2023, 23:13:24) [GCC 11.2.0]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA TITAN RTX
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.0, V11.0.194
GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 1.8.2
PyTorch compiling details: PyTorch built with:

    GCC 7.3
    C++ Version: 201402
    Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
    Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
    OpenMP 201511 (a.k.a. OpenMP 4.5)
    NNPACK is enabled
    CPU capability usage: AVX2
    CUDA Runtime 11.1
    NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
    CuDNN 8.0.5
    Magma 2.5.2
    Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.2, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON,

TorchVision: 0.9.2
OpenCV: 4.7.0
MMEngine: 0.6.0
MMCV: 2.0.0rc4
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 11.1
MMSelfSup: 1.0.0rc6+

### Describe the bug

A trained SimCLR model shows a different accuracy in linear probing with each run.
Ideally, a saved model should give same accuracy with each run. 

### Reproduces the problem - code sample

# Edited config file:
``` python

_base_ = [
    '../_base_/models/resnet50.py',
    '../_base_/datasets/custom.py',
    '../_base_/schedules/lars_coslr-90e.py',
    'mmcls::_base_/default_runtime.py',
]

model = dict(
	data_preprocessor=dict(num_classes=4), 
        backbone=dict(frozen_stages=4), 
        head=dict(num_classes=4, topk=(1,2)))


train_dataloader = dict(batch_size=512)

default_hooks = dict(
    checkpoint=dict(type='CheckpointHook', interval=10, max_keep_ckpts=3),
    #visualization=dict(type="VisualizationHook", enable=True, out_dir='working_dir/simclr_edites_edges/Custom/out_dir', interval=1)
    )
```

## used custom.py datafile

``` python

dataset_type = 'CustomDataset'
data_root = 'data/Custom/'
data_preprocessor = dict(
    num_classes=4,
    mean=[123.675, 116.28, 103.53],
    std=[58.395, 57.12, 57.375],
    to_rgb=True,
)

file_client_args = dict(backend='disk')
train_pipeline = [
    dict(type='LoadImageFromFile', file_client_args=file_client_args),
    dict(type='RandomResizedCrop', scale=224, backend='pillow'),
    dict(type='RandomFlip', prob=0.5, direction='horizontal'),
    dict(type='PackClsInputs'),
]
test_pipeline = [
    dict(type='LoadImageFromFile', file_client_args=file_client_args),
    dict(type='ResizeEdge', scale=224, edge='short', backend='pillow'),
    dict(type='CenterCrop', crop_size=224),
    dict(type='PackClsInputs'),
]

train_dataloader = dict(
    batch_size=32,
    num_workers=4,
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='meta/train.txt',
        data_prefix='train/1',
        pipeline=train_pipeline),
    sampler=dict(type='DefaultSampler', shuffle=True),
    collate_fn=dict(type='default_collate'),
    persistent_workers=True,
    pin_memory=True,
)

val_dataloader = dict(
    batch_size=32,
    num_workers=4,
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='meta/val.txt',
        data_prefix='val',
        pipeline=test_pipeline),
    sampler=dict(type='DefaultSampler', shuffle=False),
    persistent_workers=True,
)
val_evaluator = dict(type='Accuracy', topk=(1, 2))

test_dataloader = val_dataloader
test_evaluator = val_evaluator

```

### Reproduces the problem - command or script

``` python
bash ./tools/benchmarks/classification/mim_dist_test.sh \
    configs/benchmarks/classification/custom/resnet50_linear-8xb512-coslr-90e_custom.py \
    working_dir/simclr/Custom/epoch_200.pth

```

### Reproduces the problem - error message

There is no error message as code runs successfully but here is the accuracy for multiple runs of the same code using same trained model.

## Run1
03/24 13:36:35 - mmengine - INFO - Load checkpoint from working_dir/simclr_edites_edges/Custom/epoch_200.pth
03/24 13:36:36 - mmengine - INFO - Epoch(test) [24/24]  accuracy/top1: 45.4902  accuracy/top2: 80.5229

## Run2
03/24 13:50:30 - mmengine - INFO - Load checkpoint from working_dir/simclr_edites_edges/Custom/epoch_200.pth
03/24 13:50:32 - mmengine - INFO - Epoch(test) [24/24]  accuracy/top1: 20.1307  accuracy/top2: 46.7974

## Run3
03/24 13:51:04 - mmengine - INFO - Load checkpoint from working_dir/simclr_edites_edges/Custom/epoch_200.pth
03/24 13:51:06 - mmengine - INFO - Epoch(test) [24/24]  accuracy/top1: 15.4248  accuracy/top2: 37.5163

### Additional information

I expect model to return same accuracy doesn't matter the number of times I run it for.

**Comment by fangyixiao18** [comment]
did you run the test with only 1 gpu?

**Comment by guneetmutreja** [comment]
@fangyixiao18 Yes, I tested it on 1 GPU.
Apologies for delayed response.

**Comment by fangyixiao18** [comment]
> @fangyixiao18 Yes, I tested it on 1 GPU. Apologies for delayed response.

It seems you just use pretrained simclr model but didn't train the head in classification task. Thus, the classification head is random initialized, you will get ramdom results

**Comment by guneetmutreja** [comment]
@fangyixiao18 I think I trained SimCLR model on my custom dataset (Please correct me if I am doing something wrong here).
## Here is the command I used for training:
``` python
!python tools/train.py configs/selfsup/simclr/simclr_resnet50_8xb64-coslr-200e_Custom_edited_edges.py --work-dir working_dir/simclr_edites_edges/Custom
```
## and here is the command I ran to do linear probing:
``` python
!bash ./tools/benchmarks/classification/mim_dist_test.sh \
    configs/benchmarks/classification/custom/resnet50_linear-8xb512-coslr-90e_custom.py \
    working_dir/simclr_edites_edges/Custom/epoch_60.pth
```

## Here is the config file I used while training (simclr_resnet50_8xb64-coslr-200e_Custom_edited_edges.py):
``` python
_base_ = [
    '../_base_/models/simclr.py',
    #'../_base_/datasets/imagenet_simclr.py',
    '../_base_/schedules/lars_coslr-200e_in1k.py',
    '../_base_/default_runtime.py',
]

# ------------------------- dataset file---------------------------

dataset_type = 'mmcls.CustomDataset'
data_root = 'data/Custom/'
file_client_args = dict(backend='disk')

load_from = 'pretrained_models/simclr_resnet50_16xb256-coslr-800e_in1k_20220825-85fcc4de.pth'

view_pipeline = [
    #dict(type='RandomResizedCrop', size=224, backend='pillow'),
    dict(type='Resize', scale= 224, backend='pillow'),
    dict(type='RandomFlip', prob=0.5),
    dict(
        type='RandomApply',
        transforms=[
            dict(
                type='ColorJitter',
                brightness=0.8,
                contrast=0.8,
                saturation=0.8,
                hue=0.2)
        ],
        prob=0.8),
    dict(
        type='RandomGrayscale',
        prob=0.2,
        keep_channels=True,
        channel_weights=(0.114, 0.587, 0.2989)),
    dict(type='RandomGaussianBlur', sigma_min=0.1, sigma_max=2.0, prob=0.5),
]

view_pipeline2 = [
    #dict(type='RandomResizedCrop', size=224, backend='pillow'),
    dict(type='Resize', scale= 224, backend='pillow'),
    dict(type='RandomFlip', prob=0.5),
    dict(type='RandomGaussianBlur', sigma_min=0.1, sigma_max=2.0, prob=0.5),
]

train_pipeline = [
    dict(type='LoadImageFromFile', file_client_args=file_client_args),
    dict(type='MultiView', num_views=2, transforms=[view_pipeline]),
    dict(type='PackSelfSupInputs', meta_keys=['img_path'])
]

train_dataloader = dict(
    batch_size=64,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    collate_fn=dict(type='default_collate'),
    dataset=dict(
        type=dataset_type,
        data_root=data_root,
        ann_file='meta/train.txt',
        data_prefix=dict(img_path='train/1/'),
        pipeline=train_pipeline))
       
       
# ------------------------ dataset file ends -----------------------

       
# optimizer
optimizer = dict(type='LARS', lr=0.3, momentum=0.9, weight_decay=1e-6)
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=optimizer,
    paramwise_cfg=dict(
        custom_keys={
            'bn': dict(decay_mult=0, lars_exclude=True),
            'bias': dict(decay_mult=0, lars_exclude=True),
            # bn layer in ResNet block downsample module
            'downsample.1': dict(decay_mult=0, lars_exclude=True),
        }))

# runtime settings
default_hooks = dict(
    # only keeps the latest 3 checkpoints
    checkpoint=dict(type='CheckpointHook', interval=10, max_keep_ckpts=3))
```

## Here is the config file I used while linear probing (resnet50_linear-8xb512-coslr-90e_custom.py):
``` python
_base_ = [
    '../_base_/models/resnet50.py',
    '../_base_/datasets/custom.py',
    '../_base_/schedules/lars_coslr-90e.py',
    'mmcls::_base_/default_runtime.py',
]
# SimSiam linear evaluation setting
# According to SimSiam paper, this setting can also be used to evaluate
# other methods like SimCLR, MoCo, BYOL, SwAV

model = dict(
data_preprocessor=dict(num_classes=4),
        backbone=dict(frozen_stages=4),
        head=dict(num_classes=4, topk=(1,2)))

#model = dict(backbone=dict(frozen_stages=4))

# dataset summary
train_dataloader = dict(batch_size=512)

# runtime settings
default_hooks = dict(
    checkpoint=dict(type='CheckpointHook', interval=10, max_keep_ckpts=3),
    #visualization=dict(type="VisualizationHook", enable=True, out_dir='working_dir/simclr_edites_edges/Custom/out_dir', interval=1)
    )
```

**Comment by guneetmutreja** [comment]
@fangyixiao18 Did you get a chance to look at the commands I provided in previous comment?

**Comment by fangyixiao18** [comment]
The step to apply self-supervised learning is like this:
1. self-supervised learning pre-training, to obtain a pre-trained model (like backbone)
2. apply the pre-trained model, to train it on downstream tasks (like linear probing)
3. test the linear probing model

from your description, I think you may miss step 2.

**Comment by guneetmutreja** [comment]
@fangyixiao18 Super thanks for the information. I could do it now.
Closing the issue.
